File: Array&String/available_captures_for_rook.py
class Solution:
    def numRookCaptures(self, board: list[list[str]]) -> int:
        ans = 0
        def search(board):
            for i in range(8):
                for j in range(8):
                    if board[i][j] == 'R':
                        return [i, j]
        i, j = search(board)
        hl, hr = i - 1, i + 1
        if hl >= 0:
            while hl >= 0:
                if board[hl][j] == 'B' : break
                if board[hl][j] == 'p' :
                    ans += 1
                    break
                hl -= 1

        if hr <= 7:
            while hr <= 7:
                if board[hr][j] == 'B' : break
                elif board[hr][j] == 'p' :
                    ans += 1
                    break
                hr += 1

        vd, vu = j - 1, j + 1
        if vd >= 0:
            while vd >= 0:
                if board[i][vd] == 'B' : break
                elif board[i][vd] == 'p' :
                    ans += 1
                    break
                vd -= 1

        if vu <= 7:
            while vu <= 7:
                if board[i][vu] == 'B' : break
                elif board[i][vu] == 'p' :
                    ans += 1
                    break
                vu += 1
        return ans

File: Array&String/CPP/available_captures_for_rook.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int numRookCaptures(vector<vector<char>>& board) {
        vector<int> pos = search(board);
        int x0 = pos[0], y0 = pos[1], ans = 0;
        vector<vector<int>> moves = {{1,0}, {0,1}, {-1,0}, {0, -1}};
        for (auto move : moves) {
            int x = x0 + move[0];
            int y = y0 + move[1];
            while (x >= 0 && x < 8 && y >= 0 && y < 8) {
                if (board[x][y] == 'p') ans++;
                if (board[x][y] != '.') {
                    break;
                }
                x += move[0];
                y += move[1];
            }
        }
        return ans;
    }

    vector<int> search(vector<vector<char>>& board) {
        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[0].size(); j++) {
                if (board[i][j] == 'R') return vector<int>{i, j};
            }
        }
        return vector<int>{};
    }
};